plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/AppEngineConfigFileSchemaLocationProvider.java

```diff
@@ -0,0 +1,49 @@
+
+package com.google.cloud.tools.eclipse.appengine.validation;
+
+import java.net.URI;
+import java.util.Collections;
+import java.util.Map;
+import org.eclipse.core.runtime.Path;
+import org.eclipse.wst.xml.core.contentmodel.modelquery.IExternalSchemaLocationProvider;
+
+/**
+ * Associate schemas for App Engine standard environment configuration files.
+ * 
+ * @seealso the {@code org.eclipse.wst.xml.core.externalSchemaLocations} extension point
+ */
+public class AppEngineConfigFileSchemaLocationProvider implements IExternalSchemaLocationProvider {
+  // Relevant XSDs are in this bundle
+  private static final String SCHEMA_LOCATIONS_PREFIX =
+      "platform:/plugin/com.google.cloud.tools.eclipse.appengine.validation/xsd/";
+
+  @Override
+  public Map<String, String> getExternalSchemaLocation(URI fileURI) {
```

Comment:
It is almost impossible for a 0-segment fileURI since this is used for XML files.  But I'll check.

The WTP `XMLValidator` uses both the schema-location and no-namespace-schema-location.  The `XMLAssociationProvider` , which seems to be used by the autocomplete, only uses the no-namespace-schema-location.  The `CMDocumentLoader`, which is quite heavily used, uses the schema-location if the document has namespaces, or the no-namespace-schema-location otherwise.

And basically our cron.xml, datastore-indexes.xml, dispatch.xml, dos.xml, and queue.xml files don't have namespaces anyways.

Although the callers all support being returned `null`, it's not documented in the javadoc, so I'll return an empty map.

Reply:
I see. That said, `appengine-web.xml` is returned with the no-namespace-schema-location. I just wonder if it should have namespaces, unlike others.

Reply:
Never mind. I think I get it now.